Short-circuit fails to evaluate right part

This program

```xtend
			object liberarAFiona {
				var cantidadTrolls = 0
				var solicitante = ""
			
				method solicitante() = return solicitante
				method solicitante(_solicitante) { solicitante = _solicitante }
				method cantidadTrolls(cant) { cantidadTrolls = cant }
				method esDificil() {
					const result = (cantidadTrolls > 3) and (cantidadTrolls < 6)
					console.println(result)
					return result
				}
				method puntosRecompensa() = return cantidadTrolls * 2
			}
			program a {
				liberarAFiona.cantidadTrolls(5)
				liberarAFiona.esDificil()
			}
```

Fails with 

```java
java.lang.AssertionError: Error evaluating line __synthetic0.wpgm:[9]: (cantidadTrolls < 6)
wollok.lang.Exception: Error evaluating line __synthetic0.wpgm:[9]: (cantidadTrolls < 6)
	at __synthetic0.liberarAFiona.esDificil() [__synthetic0.wpgm]
	at __synthetic0.liberarAFiona.esDificil() [__synthetic0.wpgm]
	at  [__synthetic0.wpgm]

	at org.junit.Assert.fail(Assert.java:88)
	at org.uqbar.project.wollok.tests.interpreter.AbstractWollokInterpreterTestCase.lambda$7(AbstractWollokInterpreterTestCase.java:173)
	at org.uqbar.project.wollok.tests.interpreter.AbstractWollokInterpreterTestCase$$Lambda$42/1103234208.accept(Unknown Source)
	at java.lang.Iterable.forEach(Iterable.java:75)
	at org.uqbar.project.wollok.tests.interpreter.AbstractWollokInterpreterTestCase.lambda$5(AbstractWollokInterpreterTestCase.java:180)
	at org.uqbar.project.wollok.tests.interpreter.AbstractWollokInterpreterTestCase$$Lambda$2/1123236701.apply(Unknown Source)
	at org.eclipse.xtext.xbase.lib.ObjectExtensions.operator_doubleArrow(ObjectExtensions.java:139)
	at org.uqbar.project.wollok.tests.interpreter.AbstractWollokInterpreterTestCase.interpret(AbstractWollokInterpreterTestCase.java:182)
	at org.uqbar.project.wollok.tests.interpreter.AbstractWollokInterpreterTestCase.interpret(AbstractWollokInterpreterTestCase.java:138)
	at org.uqbar.project.wollok.tests.interpreter.AbstractWollokInterpreterTestCase.interpret(AbstractWollokInterpreterTestCase.java:134)
	at org.uqbar.project.wollok.tests.interpreter.AbstractWollokInterpreterTestCase.interpretPropagatingErrors(AbstractWollokInterpreterTestCase.java:102)
	at org.uqbar.project.wollok.tests.interpreter.AbstractWollokInterpreterTestCase.interpretPropagatingErrors(AbstractWollokInterpreterTestCase.java:90)
	at org.uqbar.project.wollok.tests.interpreter.BooleanTestCase.andShortcircuitBug(BooleanTestCase.java:247)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:483)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.eclipse.xtext.junit4.XtextRunner$1.evaluate(XtextRunner.java:49)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:86)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:38)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:459)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:675)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:382)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:192)

```

--- a/org.uqbar.project.wollok.lib/src/wollok/lang/AbstractJavaWrapper.xtend
+++ b/org.uqbar.project.wollok.lib/src/wollok/lang/AbstractJavaWrapper.xtend
@@ -35,7 +35,7 @@ class AbstractJavaWrapper<T> implements JavaWrapper<T> {
 	}
 	
 	def newInstanceWithWrapped(T wrapped) {
-		val transformedClassName = DefaultNativeObjectFactory.javaToWollokFQN(class.name)
+		val String transformedClassName = DefaultNativeObjectFactory.javaToWollokFQN(class.name)
 		evaluator.newInstanceWithWrapped(transformedClassName, wrapped)
 	} 
 }

--- a/org.uqbar.project.wollok.tests/src/org/uqbar/project/wollok/tests/interpreter/BooleanTestCase.xtend
+++ b/org.uqbar.project.wollok.tests/src/org/uqbar/project/wollok/tests/interpreter/BooleanTestCase.xtend
@@ -91,6 +91,30 @@ class BooleanTestCase extends AbstractWollokInterpreterTestCase {
 	}
 	
 	@Test
+	def void lazyPartOfTheAndShortCircuitAccessingTheContext() {
+		'''
+			object liberarAFiona {
+				var cantidadTrolls = 0
+				var solicitante = ""
+			
+				method solicitante() = return solicitante
+				method solicitante(_solicitante) { solicitante = _solicitante }
+				method cantidadTrolls(cant) { cantidadTrolls = cant }
+				method esDificil() {
+					const result = (cantidadTrolls > 3) and (cantidadTrolls < 6)
+					console.println(result)
+					return result
+				}
+				method puntosRecompensa() = return cantidadTrolls * 2
+			}
+			program a {
+				liberarAFiona.cantidadTrolls(5)
+				liberarAFiona.esDificil()
+			}
+		'''.interpretPropagatingErrors
+	}
+	
+	@Test
 	def void orShortcirtuitMustEvaluateSecondPart() {
 		'''
 			object p {

--- a/org.uqbar.project.wollok/src/org/uqbar/project/wollok/interpreter/WollokInterpreterEvaluator.xtend
+++ b/org.uqbar.project.wollok/src/org/uqbar/project/wollok/interpreter/WollokInterpreterEvaluator.xtend
@@ -334,7 +334,10 @@ class WollokInterpreterEvaluator implements XInterpreterEvaluator<WollokObject> 
 	}
 
 	def lazyEval(EObject expression) {
-		[| expression.eval ]
+		val lazyContext = interpreter.currentContext
+		return [| 
+			interpreter.performOnStack(expression, lazyContext, [| expression.eval])
+		]
 	}
 
 	def dispatch evaluate(WPostfixOperation op) {

